File: WebSocket/src/50/websocket_microservices_gateway.cpp
// WebSocket Gateway with Redis Integration for Microservices
// Compile: g++ -std=c++17 ws_gateway.cpp -lwebsockets -lhiredis -lpthread -o ws_gateway

#include <libwebsockets.h>
#include <hiredis/hiredis.h>
#include <string>
#include <map>
#include <vector>
#include <mutex>
#include <thread>
#include <json/json.h>
#include <iostream>

// Connection session data
struct SessionData {
    std::string user_id;
    std::string session_id;
    std::vector<std::string> subscriptions;
    lws* wsi;
};

// Global state
class GatewayState {
public:
    std::map<lws*, SessionData> sessions;
    std::mutex sessions_mutex;
    redisContext* redis_ctx;
    redisContext* redis_sub_ctx;
    bool running = true;
    
    void add_session(lws* wsi, const std::string& user_id) {
        std::lock_guard<std::mutex> lock(sessions_mutex);
        sessions[wsi] = {user_id, generate_session_id(), {}, wsi};
    }
    
    void remove_session(lws* wsi) {
        std::lock_guard<std::mutex> lock(sessions_mutex);
        sessions.erase(wsi);
    }
    
    void subscribe_to_channel(lws* wsi, const std::string& channel) {
        std::lock_guard<std::mutex> lock(sessions_mutex);
        auto it = sessions.find(wsi);
        if (it != sessions.end()) {
            it->second.subscriptions.push_back(channel);
        }
    }
    
private:
    std::string generate_session_id() {
        static int counter = 0;
        return "session_" + std::to_string(++counter);
    }
};

GatewayState gateway_state;

// Redis publisher - sends messages to microservices
class RedisPublisher {
public:
    RedisPublisher(const std::string& host, int port) {
        ctx = redisConnect(host.c_str(), port);
        if (ctx == nullptr || ctx->err) {
            throw std::runtime_error("Redis connection failed");
        }
    }
    
    ~RedisPublisher() {
        if (ctx) redisFree(ctx);
    }
    
    bool publish(const std::string& channel, const std::string& message) {
        redisReply* reply = (redisReply*)redisCommand(ctx, 
            "PUBLISH %s %s", channel.c_str(), message.c_str());
        if (reply == nullptr) return false;
        bool success = reply->type != REDIS_REPLY_ERROR;
        freeReplyObject(reply);
        return success;
    }
    
    // Call a microservice via Redis request/response pattern
    std::string call_service(const std::string& service, const std::string& request) {
        std::string response_channel = "response_" + std::to_string(time(nullptr));
        
        // Subscribe to response channel
        redisReply* reply = (redisReply*)redisCommand(ctx,
            "SUBSCRIBE %s", response_channel.c_str());
        freeReplyObject(reply);
        
        // Publish request
        Json::Value req;
        req["response_channel"] = response_channel;
        req["payload"] = request;
        publish("service." + service, req.toStyledString());
        
        // Wait for response (simplified - real implementation needs timeout)
        redisReply* msg = (redisReply*)redisCommand(ctx, "BLPOP %s 5", 
            response_channel.c_str());
        std::string response;
        if (msg && msg->type == REDIS_REPLY_ARRAY && msg->elements > 1) {
            response = msg->element[1]->str;
        }
        freeReplyObject(msg);
        
        return response;
    }
    
private:
    redisContext* ctx;
};

// Redis subscriber - receives events from microservices
void redis_subscriber_thread(const std::string& host, int port) {
    redisContext* ctx = redisConnect(host.c_str(), port);
    if (ctx == nullptr || ctx->err) {
        std::cerr << "Redis subscriber connection failed" << std::endl;
        return;
    }
    
    // Subscribe to all microservice events
    redisReply* reply = (redisReply*)redisCommand(ctx, 
        "PSUBSCRIBE events.*");
    freeReplyObject(reply);
    
    while (gateway_state.running) {
        redisReply* msg;
        if (redisGetReply(ctx, (void**)&msg) == REDIS_OK) {
            if (msg->type == REDIS_REPLY_ARRAY && msg->elements >= 4) {
                std::string channel = msg->element[2]->str;
                std::string message = msg->element[3]->str;
                
                // Parse message and route to appropriate WebSocket clients
                Json::Value event;
                Json::Reader reader;
                if (reader.parse(message, event)) {
                    // Broadcast to subscribed clients
                    std::lock_guard<std::mutex> lock(gateway_state.sessions_mutex);
                    for (auto& [wsi, session] : gateway_state.sessions) {
                        bool subscribed = false;
                        for (const auto& sub : session.subscriptions) {
                            if (channel.find(sub) != std::string::npos) {
                                subscribed = true;
                                break;
                            }
                        }
                        
                        if (subscribed) {
                            // Signal LWS to send data
                            lws_callback_on_writable(wsi);
                        }
                    }
                }
            }
            freeReplyObject(msg);
        }
    }
    
    redisFree(ctx);
}

// WebSocket protocol callback
static int callback_websocket(struct lws* wsi, enum lws_callback_reasons reason,
                              void* user, void* in, size_t len) {
    static RedisPublisher redis_pub("127.0.0.1", 6379);
    
    switch (reason) {
        case LWS_CALLBACK_ESTABLISHED: {
            std::cout << "Client connected" << std::endl;
            gateway_state.add_session(wsi, "user_" + std::to_string(time(nullptr)));
            break;
        }
        
        case LWS_CALLBACK_RECEIVE: {
            std::string message((char*)in, len);
            Json::Value msg;
            Json::Reader reader;
            
            if (reader.parse(message, msg)) {
                std::string action = msg["action"].asString();
                
                if (action == "subscribe") {
                    // Subscribe to event channels
                    std::string channel = msg["channel"].asString();
                    gateway_state.subscribe_to_channel(wsi, channel);
                    
                } else if (action == "call_service") {
                    // Make synchronous call to microservice
                    std::string service = msg["service"].asString();
                    std::string request = msg["request"].toStyledString();
                    
                    std::string response = redis_pub.call_service(service, request);
                    
                    // Send response back to client
                    Json::Value resp;
                    resp["type"] = "service_response";
                    resp["data"] = response;
                    std::string resp_str = resp.toStyledString();
                    
                    unsigned char buf[LWS_PRE + resp_str.length()];
                    memcpy(&buf[LWS_PRE], resp_str.c_str(), resp_str.length());
                    lws_write(wsi, &buf[LWS_PRE], resp_str.length(), LWS_WRITE_TEXT);
                    
                } else if (action == "publish_event") {
                    // Publish event to microservices
                    std::string channel = msg["channel"].asString();
                    std::string event = msg["event"].toStyledString();
                    redis_pub.publish("events." + channel, event);
                }
            }
            break;
        }
        
        case LWS_CALLBACK_CLOSED: {
            std::cout << "Client disconnected" << std::endl;
            gateway_state.remove_session(wsi);
            break;
        }
        
        default:
            break;
    }
    
    return 0;
}

static struct lws_protocols protocols[] = {
    {
        "ws-gateway-protocol",
        callback_websocket,
        0,
        4096,
    },
    { NULL, NULL, 0, 0 }
};

int main() {
    struct lws_context_creation_info info;
    memset(&info, 0, sizeof(info));
    
    info.port = 8080;
    info.protocols = protocols;
    info.options = LWS_SERVER_OPTION_VALIDATE_UTF8;
    
    struct lws_context* context = lws_create_context(&info);
    if (!context) {
        std::cerr << "Failed to create WebSocket context" << std::endl;
        return 1;
    }
    
    std::cout << "WebSocket Gateway running on port 8080" << std::endl;
    std::cout << "Integrating with microservices via Redis" << std::endl;
    
    // Start Redis subscriber thread
    std::thread redis_thread(redis_subscriber_thread, "127.0.0.1", 6379);
    
    // Main event loop
    while (gateway_state.running) {
        lws_service(context, 50);
    }
    
    lws_context_destroy(context);
    redis_thread.join();
    
    return 0;
}

File: WebSocket/src/50/websocket_microservices_gateway.rs
// WebSocket Gateway with Microservices Integration (Rust)
// Cargo.toml dependencies:
// tokio = { version = "1", features = ["full"] }
// tokio-tungstenite = "0.20"
// futures-util = "0.3"
// redis = { version = "0.23", features = ["tokio-comp", "connection-manager"] }
// serde = { version = "1", features = ["derive"] }
// serde_json = "1"
// tonic = "0.10"
// uuid = { version = "1", features = ["v4"] }

use tokio::net::{TcpListener, TcpStream};
use tokio_tungstenite::{accept_async, tungstenite::Message};
use futures_util::{StreamExt, SinkExt};
use redis::{Client, AsyncCommands, aio::ConnectionManager};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{RwLock, mpsc};
use uuid::Uuid;

#[derive(Clone, Debug, Serialize, Deserialize)]
struct ClientMessage {
    action: String,
    #[serde(default)]
    channel: Option<String>,
    #[serde(default)]
    service: Option<String>,
    #[serde(default)]
    request: Option<serde_json::Value>,
    #[serde(default)]
    event: Option<serde_json::Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
struct ServerMessage {
    msg_type: String,
    data: serde_json::Value,
}

#[derive(Clone)]
struct Session {
    id: String,
    user_id: String,
    subscriptions: Vec<String>,
    tx: mpsc::UnboundedSender<Message>,
}

type Sessions = Arc<RwLock<HashMap<String, Session>>>;

// Microservices communication layer
struct MicroservicesBridge {
    redis_conn: ConnectionManager,
    sessions: Sessions,
}

impl MicroservicesBridge {
    async fn new(redis_url: &str, sessions: Sessions) -> Result<Self, redis::RedisError> {
        let client = Client::open(redis_url)?;
        let redis_conn = ConnectionManager::new(client).await?;
        
        Ok(Self {
            redis_conn,
            sessions,
        })
    }
    
    // Publish event to microservices
    async fn publish_event(&mut self, channel: &str, event: &serde_json::Value) 
        -> Result<(), redis::RedisError> {
        let event_str = serde_json::to_string(event).unwrap();
        let full_channel = format!("events.{}", channel);
        
        self.redis_conn.publish(&full_channel, event_str).await?;
        println!("Published event to channel: {}", full_channel);
        
        Ok(())
    }
    
    // Call microservice using request/response pattern
    async fn call_service(&mut self, service: &str, request: &serde_json::Value) 
        -> Result<serde_json::Value, Box<dyn std::error::Error>> {
        let response_channel = format!("response_{}", Uuid::new_v4());
        let request_channel = format!("service.{}", service);
        
        // Create request with response channel
        let req = serde_json::json!({
            "response_channel": response_channel,
            "payload": request
        });
        
        // Publish request
        let req_str = serde_json::to_string(&req)?;
        self.redis_conn.publish(&request_channel, req_str).await?;
        
        // Wait for response (with timeout)
        let response: Vec<String> = tokio::time::timeout(
            std::time::Duration::from_secs(5),
            self.redis_conn.blpop(&response_channel, 0.0)
        ).await???;
        
        if response.len() > 1 {
            Ok(serde_json::from_str(&response[1])?)
        } else {
            Err("No response from service".into())
        }
    }
    
    // Subscribe to microservice events and broadcast to WebSocket clients
    async fn start_event_subscriber(sessions: Sessions, redis_url: String) {
        tokio::spawn(async move {
            let client = Client::open(redis_url.as_str()).unwrap();
            let mut conn = client.get_async_connection().await.unwrap();
            let mut pubsub = conn.into_pubsub();
            
            // Subscribe to all event channels
            pubsub.psubscribe("events.*").await.unwrap();
            println!("Subscribed to microservice events");
            
            let mut stream = pubsub.on_message();
            
            while let Some(msg) = stream.next().await {
                let channel: String = msg.get_channel_name().to_string();
                let payload: String = msg.get_payload().unwrap();
                
                if let Ok(event) = serde_json::from_str::<serde_json::Value>(&payload) {
                    // Broadcast to subscribed clients
                    let sessions = sessions.read().await;
                    
                    for (_, session) in sessions.iter() {
                        // Check if client is subscribed to this channel
                        let subscribed = session.subscriptions.iter()
                            .any(|sub| channel.contains(sub));
                        
                        if subscribed {
                            let server_msg = ServerMessage {
                                msg_type: "event".to_string(),
                                data: event.clone(),
                            };
                            
                            let msg_json = serde_json::to_string(&server_msg).unwrap();
                            let _ = session.tx.send(Message::Text(msg_json));
                        }
                    }
                }
            }
        });
    }
}

// Handle individual WebSocket connection
async fn handle_connection(
    stream: TcpStream,
    sessions: Sessions,
    redis_url: String,
) {
    let ws_stream = match accept_async(stream).await {
        Ok(ws) => ws,
        Err(e) => {
            eprintln!("WebSocket handshake failed: {}", e);
            return;
        }
    };
    
    println!("New WebSocket connection established");
    
    let (mut ws_sender, mut ws_receiver) = ws_stream.split();
    let (tx, mut rx) = mpsc::unbounded_channel();
    
    // Create session
    let session_id = Uuid::new_v4().to_string();
    let session = Session {
        id: session_id.clone(),
        user_id: format!("user_{}", Uuid::new_v4()),
        subscriptions: Vec::new(),
        tx: tx.clone(),
    };
    
    sessions.write().await.insert(session_id.clone(), session.clone());
    
    // Create microservices bridge
    let mut bridge = MicroservicesBridge::new(&redis_url, sessions.clone())
        .await
        .unwrap();
    
    // Spawn task to send messages to WebSocket client
    tokio::spawn(async move {
        while let Some(msg) = rx.recv().await {
            if ws_sender.send(msg).await.is_err() {
                break;
            }
        }
    });
    
    // Handle incoming WebSocket messages
    while let Some(msg) = ws_receiver.next().await {
        match msg {
            Ok(Message::Text(text)) => {
                if let Ok(client_msg) = serde_json::from_str::<ClientMessage>(&text) {
                    match client_msg.action.as_str() {
                        "subscribe" => {
                            if let Some(channel) = client_msg.channel {
                                let mut sessions_write = sessions.write().await;
                                if let Some(session) = sessions_write.get_mut(&session_id) {
                                    session.subscriptions.push(channel.clone());
                                    println!("Client subscribed to: {}", channel);
                                }
                            }
                        },
                        
                        "call_service" => {
                            if let (Some(service), Some(request)) = 
                                (client_msg.service, client_msg.request) {
                                match bridge.call_service(&service, &request).await {
                                    Ok(response) => {
                                        let server_msg = ServerMessage {
                                            msg_type: "service_response".to_string(),
                                            data: response,
                                        };
                                        let msg_json = serde_json::to_string(&server_msg).unwrap();
                                        let _ = tx.send(Message::Text(msg_json));
                                    },
                                    Err(e) => {
                                        eprintln!("Service call failed: {}", e);
                                    }
                                }
                            }
                        },
                        
                        "publish_event" => {
                            if let (Some(channel), Some(event)) = 
                                (client_msg.channel, client_msg.event) {
                                let _ = bridge.publish_event(&channel, &event).await;
                            }
                        },
                        
                        _ => {
                            println!("Unknown action: {}", client_msg.action);
                        }
                    }
                }
            },
            
            Ok(Message::Close(_)) => {
                println!("Client closed connection");
                break;
            },
            
            Err(e) => {
                eprintln!("WebSocket error: {}", e);
                break;
            },
            
            _ => {}
        }
    }
    
    // Clean up session
    sessions.write().await.remove(&session_id);
    println!("Session removed: {}", session_id);
}

#[tokio::main]
async fn main() {
    let addr = "127.0.0.1:8080";
    let redis_url = "redis://127.0.0.1:6379";
    
    let listener = TcpListener::bind(addr).await.unwrap();
    println!("WebSocket Gateway listening on: {}", addr);
    println!("Connecting to Redis at: {}", redis_url);
    
    let sessions: Sessions = Arc::new(RwLock::new(HashMap::new()));
    
    // Start event subscriber
    MicroservicesBridge::start_event_subscriber(
        sessions.clone(),
        redis_url.to_string()
    ).await;
    
    // Accept WebSocket connections
    while let Ok((stream, addr)) = listener.accept().await {
        println!("New connection from: {}", addr);
        
        let sessions = sessions.clone();
        let redis_url = redis_url.to_string();
        
        tokio::spawn(handle_connection(stream, sessions, redis_url));
    }
}

// Example microservice event publisher (separate process)
#[allow(dead_code)]
async fn example_microservice_publisher() {
    let client = redis::Client::open("redis://127.0.0.1:6379").unwrap();
    let mut conn = client.get_async_connection().await.unwrap();
    
    loop {
        tokio::time::sleep(std::time::Duration::from_secs(5)).await;
        
        let event = serde_json::json!({
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "event_type": "user_activity",
            "data": {
                "user_id": "12345",
                "action": "login"
            }
        });
        
        let event_str = serde_json::to_string(&event).unwrap();
        let _: () = conn.publish("events.user_activity", event_str).await.unwrap();
        
        println!("Published event from microservice");
    }
}